### PR TITLE
feat(table): add an optional vnode columns in table catalog

### LINKED ISSUE #5012: bug: state table distribution of 2-phase agg leads to wrong results
First, set the `HASH_AGG_CACHE_SIZE` to 0.

https://github.com/risingwavelabs/risingwave/blob/e8ea42c17ff383beb973d731899c2dbada4bdd1f/src/stream/src/executor/hash_agg.rs#L42-L43

Create a materialized view with 2-phase agg.

```
dev=> create table tt (v int);
CREATE_TABLE

dev=> explain create materialized view mvv as select max(v) max from tt;
                                                              QUERY PLAN                                                               
---------------------------------------------------------------------------------------------------------------------------------------
 StreamMaterialize { columns: [max], pk_columns: [] }
   StreamProject { exprs: [max(max(tt.v))] }
     StreamGlobalSimpleAgg { aggs: [sum(count), max(max(tt.v))] }
       StreamExchange { dist: Single }
         StreamHashAgg { group_key: [Vnode(tt._row_id)], aggs: [count, max(tt.v)] }
           StreamProject { exprs: [tt.v, tt._row_id, Vnode(tt._row_id)] }
             StreamTableScan { table: tt, columns: [tt.v, tt._row_id], pk: [tt._row_id], distribution: UpstreamHashShard(tt._row_id) }
(7 rows)
```

Do some inserts.
```
dev=> select * from tt;
 v 
---
 2
 1
(2 rows)

dev=> select * from mvv;
 max 
-----
   2
(1 row)
```

Flush, then delete `v = 2`.
```
dev=> delete from tt where v = 2;
DELETE 1
dev=> select * from tt;
 v 
---
 1
(1 row)

dev=> select * from mvv;
 max 
-----
   2  <- 😄
(1 row)
```

This is caused by:
- The distribution key of the local agg state store is `_row_id`, should it be singleton instead?
- The state table ignores the invalid distribution key computation, and uses the vnode `0` (why?) for reads.
- The recovery of states / deletion is not tested well.

Related:
- https://github.com/risingwavelabs/risingwave/pull/5011#issuecomment-1231926678

cc @st1page @richardchien @wcy-fdu

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
add an optional vnode column index field in HashAgg, tableCatalog and stateTable
will fix topN later.
fix #5012 
## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: proto/catalog.proto [hunk 1]
```diff
@@ -87,6 +87,8 @@ message Table {
   uint32 owner = 15;
   map<string, string> properties = 16;
   uint32 fragment_id = 17;
+  // a optional column index which is the vnode of each row computed by the table's consistent hash distribution
```
Comment (BugenZhao) [inline proto/catalog.proto:90]:
```suggestion
  // an optional column index which is the vnode of each row computed by the table's consistent
  // hash distribution, currently only used for 2-phase local aggregation.
```

File: src/frontend/src/catalog/table_catalog.rs [hunk 2]
```diff
@@ -172,6 +176,9 @@ impl TableCatalog {
             owner: self.owner,
             properties: self.properties.clone(),
             fragment_id: self.fragment_id,
+            vnode_col_idx: self
+                .vnode_col_idx
+                .map(|i| ProstColumnIndex { index: i as _ }),
```
Comment (BugenZhao) [inline src/frontend/src/catalog/table_catalog.rs:181]: Should we also use `ColumnIndex` for keys? 😄

File: src/storage/src/table/streaming_table/state_table.rs [hunk 3]
```diff
@@ -88,6 +88,10 @@ pub struct StateTable<S: StateStore> {
 
     /// If true, sanity check is disabled on this table.
     disable_sanity_check: bool,
+
+    /// a optional column index which is the vnode of each row computed by the table's consistent
+    /// hash distribution
```
Comment (BugenZhao) [inline src/storage/src/table/streaming_table/state_table.rs:93]: Ditto.

**Review by BugenZhao — APPROVED** [review]
LGTM

File: src/frontend/src/catalog/table_catalog.rs (hunk 2 above)
Comment (st1page) [reply src/frontend/src/catalog/table_catalog.rs:181]: It is just for express a option in proto

**Pull request merged**